tasks/FileUploadTask.md

```diff
@@ -15,6 +15,8 @@ This task aims to provide a fluent and easy to use mechanism for the consumer to
 - Using the RequestContext, the feature flag for the FileUploadTask can be set for telemetry purposes.
 - In the event that a slice fails to upload, the task should retry to re-upload the slice for a configurable number of times(default of 3) before giving up and throwing an error/exception for the user to handle.
 - The task should be agnostic to the kind of upload being perfromed so as to support for various fileUpload scenarios e.g. DriveItem and FileAttachment
+- An upload task should be marked as completed if the response status is a 201. Another condition valid only for OneDrive is if the response status is a 200 and the response contains an "id" then the task is complete.
```

Comment:
Does this mean that an upload task is completed for:
* file attachments when the response status is a 201
* drive items when the response status is 200 and the response contains an `id` property in the response body? 

For the OneDrive scenario, the response contains a driveitem object, not just an id.

Reply:
https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0&preserve-view=true

> If the file can be committed using the new metadata, an HTTP 201 Created or HTTP 200 OK response will be returned with the Item metadata for the uploaded file.

Drive items also have 201 response status.

Reply:
When I worked on that I noticed a different behavior between OneDrive for business and OneDrive consumer, I don't remember which behaves what way though.

The way this is now implemented in Java, is looks for the response body:
- if the response is an error, well something went wrong
- if the response is an upload session, chunk accepted, send the next bytes
- if the response is an object representation (drive, attachment...) or contains a location header, we're done

This is a bit cumbersome, but it accounts for the different implementations from the different services and utilization contexts.

Reply:
This is what I learnt -> 
[3:12 PM] Nikitha Udaykumar Chettiar
I have a question about large file uploads in OneDrive. https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0&preserve-view=true

Does this behaviour differ between OneDrive for consumer and OneDrive for Business? 

​[3:12 PM] Brad Cleaver
    no that contract should work for both
​[3:13 PM] Brad Cleaver
    the only difference you may see is that in ODB the upload of new files will generate a temp file that's visible in some views
​[3:13 PM] Brad Cleaver
    that does not occur for ODC

If there is still a difference in the behaviour then that should be a service issue.

Reply:
I agree with the sentiment that it would be fit to describe it as "response contains a driveitem object" and not just an Id.

Otherwise, this looks good!

Reply:
Printdocument file upload is modeled after OneDrive. After the last range is uploaded, the response is 201 with printdocument object populated with id, contenttype, displayname, and size.
Adding @nilakhan

Reply:
Got this information from @nilakhan. 
https://docs.microsoft.com/en-us/graph/api/printdocument-createuploadsession

https://docs.microsoft.com/en-us/graph/upload-data-to-upload-session

Reply:
I feel checking for response status as 201 to mark the upload completion would cover all cases.